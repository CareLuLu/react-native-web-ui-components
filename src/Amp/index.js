import React, { useContext } from 'react';

const AMP = React.createContext('amp');

export const { Provider } = AMP;

export const { Consumer } = AMP;

export const withAmp = () => Component => props => (
  <AMP.Consumer>
    {amp => <Component {...props} amp={amp} />}
  </AMP.Consumer>
);

export const useAmp = () => useContext(AMP);

export default {
  Provider,
  Consumer,
  useAmp,
  withAmp,
};
