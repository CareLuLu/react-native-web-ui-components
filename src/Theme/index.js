import React, { useContext } from 'react';
import { StyleSheet, Platform } from 'react-native';
import omit from 'lodash/omit';
import PropTypes from 'prop-types';
import merge from 'lodash/merge';

const Theme = React.createContext('theme');

const themeProps = [
  'fontFamily',
  'theme',
  'themeTextStyle',
  'themeInputStyle',
  'themePrimaryStyle',
];

const defaults = {
  canonical: uri => uri.replace('/amp'),
  resource: uri => uri,
  omit: props => omit(props, themeProps),
  '*': {
    fontFamily: {
      regular: 'Lucida Sans',
      bold: 'Lucida Sans Bold',
    },
  },
  Link: {
    basepath: 'localhost',
  },
  Datepicker: {
    selectedDateColor: '#337AB7',
  },
  input: {
    regular: StyleSheet.create({
      background: { backgroundColor: '#FFFFFF' },
      border: {
        borderWidth: 1,
        borderStyle: 'solid',
        borderColor: '#D3D6D6',
        borderRadius: 2,
      },
      text: { color: '#545454' },
      placeholder: { color: '#D3D6D6' },
      opacity: { opacity: 1 },
      selected: { color: '#0E73CA' },
      unselected: { color: '#BDC3C7' },
    }),
    readonly: StyleSheet.create({
      background: { backgroundColor: '#F4F6F6' },
      border: {
        borderWidth: 1,
        borderStyle: 'solid',
        borderColor: '#D5DBDB',
        borderRadius: 2,
      },
      text: { color: '#545454' },
      placeholder: { color: '#D3D6D6' },
      opacity: { opacity: 0.7 },
      selected: { color: '#0E73CA' },
      unselected: { color: '#BDC3C7' },
    }),
    disabled: StyleSheet.create({
      background: { backgroundColor: '#F4F6F6' },
      border: {
        borderWidth: 1,
        borderStyle: 'solid',
        borderColor: '#D5DBDB',
        borderRadius: 2,
      },
      text: { color: '#545454' },
      placeholder: { color: '#D3D6D6' },
      opacity: { opacity: 0.7 },
      selected: { color: '#0E73CA' },
      unselected: { color: '#BDC3C7' },
    }),
    focused: StyleSheet.create({
      background: { backgroundColor: '#FFFFFF' },
      border: {
        borderWidth: 1,
        borderStyle: 'solid',
        borderColor: '#6CB3FF',
        borderRadius: 2,
      },
      text: { color: '#545454' },
      placeholder: { color: '#D3D6D6' },
      opacity: { opacity: 1 },
      selected: { color: '#0E73CA' },
      unselected: { color: '#BDC3C7' },
    }),
    error: StyleSheet.create({
      background: { backgroundColor: '#FFFFFF' },
      border: {
        borderWidth: 1,
        borderStyle: 'solid',
        borderColor: '#EE2D68',
        borderRadius: 2,
      },
      text: { color: '#545454' },
      placeholder: { color: '#EE2D68' },
      opacity: { opacity: 1 },
      selected: { color: '#0E73CA' },
      unselected: { color: '#BDC3C7' },
    }),
  },
  colors: {
    text: 'gray',
    primary: 'navy',
    pink: StyleSheet.create({
      background: { backgroundColor: '#F15786' },
      border: { borderColor: '#BF4E71' },
      text: { color: '#F15786' },
    }),
    yellow: StyleSheet.create({
      background: { backgroundColor: '#FEB715' },
      border: { borderColor: '#D99D14' },
      text: { color: '#FEB715' },
    }),
    gray: StyleSheet.create({
      background: { backgroundColor: '#D4D4D4' },
      border: { borderColor: '#BFBFBF' },
      text: { color: '#545454' },
    }),
    lightGray: StyleSheet.create({
      background: { backgroundColor: '#BFBFBF' },
      border: { borderColor: '#BFBFBF' },
      text: { color: '#BFBFBF' },
    }),
    white: StyleSheet.create({
      background: { backgroundColor: '#FFFFFF' },
      border: { borderColor: '#BFBFBF' },
      text: { color: '#FFFFFF' },
    }),
    navy: StyleSheet.create({
      background: { backgroundColor: '#0E73CA' },
      border: { borderColor: '#055396' },
      text: { color: '#0E73CA' },
    }),
    teal: StyleSheet.create({
      background: { backgroundColor: '#23AAAA' },
      border: { borderColor: '#23AAAA' },
      text: { color: '#23AAAA' },
    }),
    black: StyleSheet.create({
      background: { backgroundColor: '#000000' },
      border: { borderColor: '#000000' },
      text: { color: '#000000' },
    }),
  },
  platform: {
    web: {
      '*': {
        fontFamily: {
          regular: '"Lucida Sans Unicode","Lucida Grande",Arial,Helvetica,clean,sans-serif',
          bold: '"Lucida Grande", "Lucida Sans Unicode","Lucida Grande",Arial,Helvetica,clean,sans-serif',
        },
      },
    },
  },
};

const getInputStyle = (theme, {
  disabled,
  readonly,
  hasError,
  autoFocus,
}) => {
  let label = 'regular';
  if (hasError) {
    label = 'error';
  } else if (disabled) {
    label = 'disabled';
  } else if (readonly) {
    label = 'readonly';
  } else if (autoFocus) {
    label = 'focused';
  }
  return theme.input[label];
};

export class Provider extends React.Component {
  static propTypes = {
    value: PropTypes.shape(),
  };

  static defaultProps = {
    value: {},
  };

  static getDerivedStateFromProps(nextProps, prevState) {
    if (nextProps.value !== prevState.value) {
      return { value: merge({}, defaults, nextProps.value) };
    }
    return null;
  }

  constructor(props) {
    super(props);
    const { value } = props;
    this.state = { value: merge({}, defaults, value) };
  }

  render() {
    const { value } = this.state;
    return <Theme.Provider {...this.props} value={value} />;
  }
}

export const { Consumer } = Theme;

export const useTheme = (type, { style, ...props }) => {
  const theme = useContext(Theme);
  return {
    ...(theme['*'] || {}),
    ...((theme.platform[Platform.OS] && theme.platform[Platform.OS]['*']) || {}),
    ...(theme[type] || {}),
    ...((theme.platform[Platform.OS] && theme.platform[Platform.OS][type]) || {}),
    ...props,
    style: [theme[type] && theme[type].style, style],
    theme,
    themeTextStyle: theme.colors[theme.colors.text],
    themePrimaryStyle: theme.colors[theme.colors.primary],
    themeInputStyle: getInputStyle(theme, props),
  };
};

export const withTheme = type => Component => ({ style, ...props }) => { // eslint-disable-line
  Component.displayName = type; // eslint-disable-line
  return (
    <Theme.Consumer>
      {theme => (
        <Component
          {...(theme['*'] || {})}
          {...((theme.platform[Platform.OS] && theme.platform[Platform.OS]['*']) || {})}
          {...(theme[type] || {})}
          {...((theme.platform[Platform.OS] && theme.platform[Platform.OS][type]) || {})}
          {...props}
          style={[theme[type] && theme[type].style, style]}
          theme={theme}
          themeTextStyle={theme.colors[theme.colors.text]}
          themePrimaryStyle={theme.colors[theme.colors.primary]}
          themeInputStyle={getInputStyle(theme, props)}
        />
      )}
    </Theme.Consumer>
  );
};

export default {
  withTheme,
  Provider,
  Consumer,
};
