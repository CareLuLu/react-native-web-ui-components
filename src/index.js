export { default as Alert } from './Alert';
export { default as Amp } from './Amp';
export { default as Banner } from './Banner';
export { default as Bold } from './Bold';
export { default as Box } from './Box';
export { default as BoxButtonRow } from './BoxButtonRow';
export { default as BoxHeader } from './BoxHeader';
export { default as BoxItem } from './BoxItem';
export { default as BoxTitle } from './BoxTitle';
export { default as Button } from './Button';
export { default as Carousel } from './Carousel';
export { default as Center } from './Center';
export { default as Checkbox } from './Checkbox';
export { default as Column } from './Column';
export { default as Confirm } from './Confirm';
export { default as Container } from './Container';
export { default as createDomStyle } from './createDomStyle';
export { default as Datepicker } from './Datepicker';
export { default as Draggable } from './Draggable';
export { default as Dropzone } from './Dropzone';
export { default as Helmet } from './Helmet';
export { default as HideShowText } from './HideShowText';
export { default as Hr } from './Hr';
export { default as Icon } from './Icon';
export { default as IconLink } from './IconLink';
export { default as IconText } from './IconText';
export { default as Image } from './Image';
export { default as Link } from './Link';
export { default as Loading } from './Loading';
export { default as MainContainer } from './MainContainer';
export { default as Modal } from './Modal';
export { default as ModalPicker } from './ModalPicker';
export { default as NavLink } from './NavLink';
export { default as Popup } from './Popup';
export { default as Radiobox } from './Radiobox';
export { default as Rating } from './Rating';
export { default as ReadMoreLessLink } from './ReadMoreLessLink';
export { default as Router } from './Router';
export { default as Row } from './Row';
export { default as Screen } from './Screen';
export { default as ScrollView } from './ScrollView';
export { default as Select } from './Select';
export { default as Sidebar } from './Sidebar';
export { default as Slider } from './Slider';
export { default as Spinner } from './Spinner';
export { default as StylePropType } from './StylePropType';
export { default as Switch } from './Switch';
export { default as Text } from './Text';
export { default as TextInput } from './TextInput';
export { default as Theme } from './Theme';
export { default as TimeRangePicker } from './TimeRangePicker';
export { default as Title } from './Title';
export { default as Tooltip } from './Tooltip';
export { default as TouchableOpacity } from './TouchableOpacity';
export { default as TouchableWithoutFeedback } from './TouchableWithoutFeedback';
export { default as UIProvider } from './UIProvider';
export { default as View } from './View';
export { default as WebOnly } from './WebOnly';
export { default as WebView } from './WebView';
