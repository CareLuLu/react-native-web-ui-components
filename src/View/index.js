import React from 'react';
import PropTypes from 'prop-types';
import { View as RNView } from 'react-native';
import noop from 'lodash/noop';
import { withTheme } from '../Theme';

const View = ({
  theme,
  onRef,
  className,
  ...props
}) => (
  <RNView ref={onRef} {...theme.omit(props)} data-class={className} />
);

View.propTypes = {
  theme: PropTypes.shape().isRequired,
  className: PropTypes.string,
  onRef: PropTypes.func,
};

View.defaultProps = {
  className: '',
  onRef: noop,
};

export default withTheme('View')(View);
