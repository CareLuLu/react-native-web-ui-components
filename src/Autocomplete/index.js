import React from 'react';
import PropTypes from 'prop-types';
import { Platform, StyleSheet } from 'react-native';
import { throttle, debounce } from 'throttle-debounce';
import noop from 'lodash/noop';
import omit from 'lodash/omit';
import filter from 'lodash/filter';
import isFunction from 'lodash/isFunction';
import isEqual from 'lodash/isEqual';
import { isEmpty } from '../utils';
import { withTheme } from '../Theme';
import Screen, { withKeyboard } from '../Screen';
import EventHandler from '../EventHandler';
import View from '../View';
import StylePropType from '../StylePropType';
import DefaultInput from '../TextInput';
import DefaultMenu from './Menu';

const defaultGetItemValue = item => item;

const defaultIsMatch = (text, item, { getItemValue }) => {
  if (text === '' || `${getItemValue(item)}`.toLowerCase().indexOf(text.toLowerCase()) >= 0) {
    return true;
  }
  return false;
};

const isField = (element, classNameRegex) => {
  for (let node = element; node && node !== document; node = node.parentNode) {
    if (classNameRegex.test(node.getAttribute('data-class') || '')) {
      return true;
    }
  }
  return false;
};

const propNames = [
  'items',
  'getItemValue',
  'onChangeText',
  'onKeyPress',
  'onSelect',
  'onSubmitEditing',
  'isMatch',
  'Input',
  'Menu',
  'menuStyle',
  'Item',
  'itemStyle',
  'itemHeight',
  'itemProps',
  'Spinner',
  'spinnerHeight',
  'EmptyResult',
  'emptyResultHeight',
  'containerStyle',
  'throttleDelay',
  'debounceDelay',
  'throttleDebounceThreshold',
];

class Autocomplete extends EventHandler {
  static propTypes = {
    keyboard: PropTypes.number.isRequired,
    items: PropTypes.oneOfType([PropTypes.array, PropTypes.func]),
    getItemValue: PropTypes.func,
    onChangeText: PropTypes.func,
    onKeyPress: PropTypes.func,
    onSelect: PropTypes.func,
    onRef: PropTypes.func,
    onSubmitEditing: PropTypes.func,
    isMatch: PropTypes.func,
    Input: PropTypes.elementType,
    Menu: PropTypes.elementType,
    menuStyle: StylePropType,
    containerStyle: StylePropType,
    name: PropTypes.string,
    value: PropTypes.any, // eslint-disable-line
    itemHeight: PropTypes.number,
    spinnerHeight: PropTypes.number,
    emptyResultHeight: PropTypes.number,
    throttleDelay: PropTypes.number,
    debounceDelay: PropTypes.number,
    throttleDebounceThreshold: PropTypes.number,
  };

  static defaultProps = {
    items: [],
    getItemValue: defaultGetItemValue,
    onChangeText: noop,
    onKeyPress: noop,
    onSelect: noop,
    onRef: noop,
    onSubmitEditing: noop,
    isMatch: defaultIsMatch,
    Input: DefaultInput,
    Menu: DefaultMenu,
    menuStyle: null,
    containerStyle: null,
    name: undefined,
    value: undefined,
    itemHeight: 30,
    spinnerHeight: 30,
    emptyResultHeight: 30,
    throttleDelay: 500,
    debounceDelay: 500,
    throttleDebounceThreshold: 3,
  };

  constructor(props) {
    super(props);
    const {
      items,
      value,
      name,
      throttleDelay,
      debounceDelay,
    } = props;
    this.fy = 0;
    this.state = {
      items: isFunction(items) ? [] : items,
      open: false,
      loading: isFunction(items),
      highlightedIndex: 0,
      keyboardOffset: null,
    };
    this.updateItemsThrottled = this.updateItems;
    this.updateItemsDebounced = this.updateItems;
    if (isFunction(items)) {
      this.updateItemsThrottled = throttle(throttleDelay, this.updateItems);
      this.updateItemsDebounced = debounce(debounceDelay, this.updateItems);
    }
    this.id = `Autocomplete__${name || Math.random().toString(36).substr(2, 9)}`;
    this.fieldRegex = new RegExp(`\\b${this.id}\\b`);
    setTimeout(() => this.updateItems(value));
  }

  componentDidMount() {
    if (Platform.OS === 'web') {
      window.addEventListener('click', this.clickListener);
    }
    this.mounted = true;
    this.onMount();
  }

  componentWillUnmount() {
    this.mounted = false;
    if (Platform.OS === 'web') {
      window.removeEventListener('click', this.clickListener);
    }
  }

  onChangeText = (text) => {
    const { onChangeText, throttleDebounceThreshold } = this.props;
    onChangeText(text);
    this.onMount(() => this.setState({
      open: true,
      loading: true,
    }));
    if (text.length > throttleDebounceThreshold) {
      this.updateItemsDebounced(text);
    } else {
      this.updateItemsThrottled(text);
    }
  };

  onKeyPress = (event) => {
    const { onKeyPress } = this.props;
    onKeyPress(event);
    if (!event.isDefaultPrevented() && Platform.OS === 'web') {
      const { key } = event.nativeEvent;
      if (key === 'Escape') {
        event.preventDefault();
        this.onMount(() => this.setState({ open: false }));
      } else if (key === 'ArrowUp' || key === 'ArrowDown') {
        const { items, highlightedIndex } = this.state;
        const increment = key === 'ArrowUp' ? -1 : 1;
        let nextHighlightedIndex = (highlightedIndex + increment) % items.length;
        if (nextHighlightedIndex < 0) {
          nextHighlightedIndex = items.length + nextHighlightedIndex;
        }
        event.preventDefault();
        this.onMount(() => this.setState({ highlightedIndex: nextHighlightedIndex }));
      }
    }
  };

  onKeyboardOpen = () => {
    const self = this;
    self.onMount(() => self.setState({ keyboardOffset: 0 }));
    if (self.input) {
      self.input.measureInWindow((x, y) => {
        const { keyboard } = self.props;
        const keyboardOffset = Screen.getHeight() - keyboard - y;
        if (keyboard > 0 && keyboardOffset < 0) {
          self.onMount(() => self.setState({ keyboardOffset }));
        }
      });
    }
  };

  onKeyboardClose = () => this.onMount(() => this.setState({ keyboardOffset: null }));

  onSubmitEditing = (event) => {
    const { open, items, highlightedIndex } = this.state;
    const { onSubmitEditing } = this.props;
    if (Platform.OS === 'web' && open && items.length > highlightedIndex) {
      this.onSelect(items[highlightedIndex], highlightedIndex);
      event.preventDefault();
    } else {
      onSubmitEditing(event);
    }
  };

  onSelect = (item, index) => {
    const { onSelect, getItemValue } = this.props;
    onSelect(getItemValue(item), item);
    this.onMount(() => this.setState({
      open: false,
      loading: false,
      highlightedIndex: index,
    }));
  };

  onRef = (input) => {
    const self = this;
    self.input = input;
    const { onRef } = self.props;
    if (isFunction(onRef)) {
      onRef(input);
    }
    setTimeout(() => {
      if (self.input) {
        self.input.measure((fx, fy, width, height, px, py) => {
          self.fy = fy;
          if (Platform.OS === 'web') {
            self.menuStyle = {
              width,
              top: height,
              left: fx,
              maxHeight: 200,
            };
          } else {
            const maxHeight = Math.max(0, Math.min(200, py));
            self.menuStyle = {
              width,
              maxHeight,
              top: fy - maxHeight,
              left: fx,
            };
          }
        });
      }
    });
  };

  clickListener = (event) => {
    if (Platform.OS === 'web') {
      if (!isField(event.target, this.fieldRegex)) {
        this.onMount(() => this.setState({ open: false }));
      }
    }
  };

  getParams() {
    return { ...this.props, ...this.state };
  }

  filterItems(text, items) {
    const { isMatch } = this.props;
    return filter(items, item => isMatch(text, item, this.getParams()));
  }

  async getItems(text) {
    let { items } = this.props;
    let parsedText = text;
    if (parsedText === null || parsedText === undefined) {
      parsedText = '';
    }
    const id = Math.random().toString(36).substr(2, 9);
    this.request = {
      id,
      loading: true,
      items: this.state.items,
    };
    if (isFunction(items)) {
      items = await items(parsedText, this.getParams());
      if (this.request.id === id) {
        this.request.loading = false;
        this.request.items = items;
      } else {
        return this.request.items;
      }
    } else {
      items = this.filterItems(parsedText, items);
      this.request.loading = false;
      this.request.items = items;
    }
    return items;
  }

  updateItems = async (text) => {
    const items = await this.getItems(text);
    const highlightedIndex = isEqual(items, this.state.items) ? this.state.highlightedIndex : 0;
    this.onMount(() => this.setState({
      items,
      highlightedIndex,
      loading: this.request.loading,
    }));
  };

  render() {
    const {
      open,
      items,
      loading,
      keyboardOffset,
    } = this.state;
    const {
      Menu,
      Input,
      style,
      itemHeight,
      spinnerHeight,
      emptyResultHeight,
      menuStyle,
      keyboard,
      containerStyle,
    } = this.props;
    const props = omit(this.props, propNames);
    if (Platform.OS === 'web') {
      props.autoComplete = 'off';
    } else {
      props.autoCompleteType = 'off';
    }
    const { minWidth, maxWidth, width } = StyleSheet.flatten([style]);
    const height = (
      2
      + (items.length * itemHeight)
      + (loading ? spinnerHeight : 0)
      + (!items.length && !loading ? emptyResultHeight : 0)
    );
    let mobileMenuStyle = null;
    if (Platform.OS !== 'web') {
      mobileMenuStyle = { top: this.fy - height + (keyboardOffset || 0) };
      if (keyboard > 0 && keyboardOffset === null) {
        setTimeout(this.onKeyboardOpen);
      }
      if (keyboard === 0 && keyboardOffset !== null) {
        setTimeout(this.onKeyboardClose);
      }
    }

    return (
      <View className={this.id} style={[{ minWidth, maxWidth, width }, containerStyle]}>
        <Input
          {...props}
          onRef={this.onRef}
          onKeyPress={this.onKeyPress}
          onChangeText={this.onChangeText}
          onSubmitEditing={this.onSubmitEditing}
        />
        {open && !isEmpty(props.value) ? (
          <Menu
            {...this.props}
            {...this.state}
            style={[
              this.menuStyle,
              { height },
              mobileMenuStyle,
              menuStyle,
            ]}
            onSelect={this.onSelect}
          />
        ) : null}
      </View>
    );
  }
}

export default withTheme('Autocomplete')(withKeyboard()(Autocomplete));
