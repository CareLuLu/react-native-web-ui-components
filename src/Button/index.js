import React from 'react';
import PropTypes from 'prop-types';
import noop from 'lodash/noop';
import { StyleSheet } from 'react-native';
import StylePropType from '../StylePropType';
import { withTheme } from '../Theme';
import Text from '../Text';
import Link from '../Link';
import View from '../View';
import TouchableOpacity from '../TouchableOpacity';

const styles = StyleSheet.create({
  empty: {},
  auto: {
    alignSelf: 'flex-start',
  },
  fullWidth: {
    width: '100%',
  },
  outer: {
    marginRight: 5,
    marginBottom: 5,
    borderBottomWidth: 2,
  },
  whiteOuter: {
    borderBottomWidth: 1,
    borderWidth: 1,
  },
  flat: {
    borderBottomWidth: 0,
    borderWidth: 0,
  },
  radius: {
    borderRadius: 2,
  },
  text: {
    textAlign: 'center',
    fontSize: 14,
    paddingTop: 7,
    paddingBottom: 7,
    paddingLeft: 20,
    paddingRight: 20,
  },
  small: {
    fontSize: 13,
    paddingTop: 5,
    paddingBottom: 5,
    paddingLeft: 10,
    paddingRight: 10,
  },
  extraSmall: {
    fontSize: 11,
    paddingTop: 2,
    paddingBottom: 2,
    paddingLeft: 7,
    paddingRight: 7,
  },
  nomargin: {
    marginRight: 0,
    marginBottom: 0,
  },
});

const getTextType = (textType, type) => {
  if (textType) {
    return textType;
  }
  if (type === 'blue') {
    return 'blue';
  }
  if (type === 'white') {
    return 'gray';
  }
  return 'white';
};

const ifProp = (props, propName) => {
  if (props[propName]) {
    return styles[propName];
  }
  return styles.empty;
};

const Button = (props) => {
  const {
    to,
    auto,
    type,
    style,
    textStyle,
    blank,
    onPress,
    replace,
    children,
    textType,
    theme,
    className,
  } = props;
  let width = null;
  let height = null;
  let isAuto = auto;
  let buttonWidth = isAuto ? 'button-auto' : 'button-full';
  const css = StyleSheet.flatten(style || {});
  if (css.width !== undefined) {
    isAuto = true;
    buttonWidth = 'button-auto';
    width = { width: css.width };
  }
  if (css.height !== undefined) {
    height = { height: css.height };
  }
  return (
    <TouchableOpacity
      className={`${className} button-wrapper ${buttonWidth}`}
      onPress={onPress}
      style={[isAuto ? styles.empty : styles.fullWidth, style]}
    >
      <View
        className={`button button-${type}`}
        style={[
          styles.outer,
          theme.colors[type] && theme.colors[type].border,
          theme.colors[type] && theme.colors[type].background,
          styles[`${type}Outer`],
          isAuto ? styles.auto : styles.empty,
          ifProp(props, 'flat'),
          ifProp(props, 'radius'),
          ifProp(props, 'nomargin'),
          width,
          height,
        ]}
      >
        {to ? (
          <Link
            replace={replace}
            type={getTextType(textType, type)}
            blank={blank}
            to={to}
            style={[styles.text, ifProp(props, 'small'), ifProp(props, 'extraSmall'), textStyle]}
          >
            {children}
          </Link>
        ) : (
          <Text
            type={getTextType(textType, type)}
            style={[styles.text, ifProp(props, 'small'), ifProp(props, 'extraSmall'), textStyle]}
          >
            {children}
          </Text>
        )}
      </View>
    </TouchableOpacity>
  );
};

Button.propTypes = {
  theme: PropTypes.shape().isRequired,
  type: PropTypes.string.isRequired,
  onPress: PropTypes.func,
  children: PropTypes.node,
  to: PropTypes.string,
  blank: PropTypes.bool,
  textType: PropTypes.string,
  style: StylePropType,
  textStyle: StylePropType,
  auto: PropTypes.bool,
  className: PropTypes.string,
  replace: PropTypes.bool,
  small: PropTypes.bool, // eslint-disable-line
  extraSmall: PropTypes.bool, // eslint-disable-line
  flat: PropTypes.bool,  // eslint-disable-line
  radius: PropTypes.bool,  // eslint-disable-line
  nomargin: PropTypes.bool,  // eslint-disable-line
};

Button.defaultProps = {
  children: null,
  small: false,
  extraSmall: false,
  flat: true,
  radius: false,
  onPress: noop,
  to: null,
  blank: false,
  textType: null,
  style: null,
  textStyle: null,
  nomargin: false,
  auto: false,
  className: '',
  replace: false,
};

export default withTheme('Button')(Button);
