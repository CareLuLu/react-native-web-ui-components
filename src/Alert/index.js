import React, { useState, useCallback } from 'react';
import PropTypes from 'prop-types';
import { StyleSheet } from 'react-native';
import noop from 'lodash/noop';
import { withTheme } from '../Theme';
import Popup from '../Popup';
import Button from '../Button';
import Center from '../Center';
import Text from '../Text';

const styles = StyleSheet.create({
  text: {
    textAlign: 'center',
    paddingTop: 5,
    paddingBottom: 15,
  },
});

const Alert = ({
  theme,
  children,
  visible,
  onOk,
  ...props
}) => {
  const [isVisible, setVisible] = useState(visible);
  const hide = useCallback(() => {
    setVisible(false);
    onOk();
  }, [setVisible, onOk]);

  return (
    <Popup
      shouldCloseOnEsc
      {...props}
      onClose={onOk}
      followKeyboard
      visible={isVisible}
      className="NoAdjustment"
    >
      {typeof children === 'string' ? (
        <Text style={styles.text}>{children}</Text>
      ) : children}
      <Center>
        <Button auto type={theme.colors.primary} flat={false} onPress={hide}>Ok</Button>
      </Center>
    </Popup>
  );
};

Alert.propTypes = {
  theme: PropTypes.shape().isRequired,
  visible: PropTypes.bool,
  children: PropTypes.node,
  onOk: PropTypes.func,
};

Alert.defaultProps = {
  visible: true,
  children: null,
  onOk: noop,
};

export default withTheme('Alert')(Alert);
