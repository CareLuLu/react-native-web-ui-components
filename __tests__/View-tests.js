import React from 'react';
import renderer from 'react-test-renderer';
import View from '../src/View';
import Button from '../src/Button';
import { Provider as ThemeProvider } from '../src/Theme';

let testDescription = '';
testDescription = 'Outer View should render with class="Test__parent class".\n'
                    + 'Inner View should render with class="Test__child".\n';
test(testDescription, () => {
  const tree = renderer.create(
    <ThemeProvider>
      <View className="Test__parent">
        <View className="Test__child" />
      </View>
    </ThemeProvider>,
  ).toJSON();
  expect(tree).toMatchSnapshot();
});

testDescription = 'View should render with class="Test__class".\n'
                    + 'There should have an onRef referencing to a Function.';
test(testDescription, () => {
  let view;
  const tree = renderer.create(
    <ThemeProvider>
      <View class="Test__class" onref={(component) => { view = component; }}>
        <Button type="navy" onPress={() => { view.setNativeProps({ style: { backgroundColor: '#20B2AA' } }); }}>
            Change Background Color
        </Button>
      </View>
    </ThemeProvider>,
  ).toJSON();
  expect(tree).toMatchSnapshot();
});

testDescription = 'View should render with class="".\n'
                    + 'There should have an onRef referencing to a Function.';
test(testDescription, () => {
  let view;
  const tree = renderer.create(
    <ThemeProvider>
      <View onref={(component) => { view = component; }}>
        <Button type="navy" onPress={() => { view.setNativeProps({ style: { backgroundColor: '#20B2AA' } }); }}>
            Change Background Color
        </Button>
      </View>
    </ThemeProvider>,
  ).toJSON();
  expect(tree).toMatchSnapshot();
});
