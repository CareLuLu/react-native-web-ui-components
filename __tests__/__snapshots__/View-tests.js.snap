// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Outer View should render with class="Test__parent class".
Inner View should render with class="Test__child".
 1`] = `
<View
  data-class="Test__parent"
  style={
    Array [
      undefined,
      undefined,
    ]
  }
>
  <View
    data-class="Test__child"
    style={
      Array [
        undefined,
        undefined,
      ]
    }
  />
</View>
`;

exports[`View should render with class="".
There should have an onRef referencing to a Function. 1`] = `
<View
  data-class=""
  onref={[Function]}
  style={
    Array [
      undefined,
      undefined,
    ]
  }
>
  <View
    accessible={true}
    isTVSelectable={true}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "opacity": 1,
        "width": "100%",
      }
    }
  >
    <View
      data-class="button button-navy"
      style={
        Array [
          undefined,
          Array [
            Object {
              "borderBottomWidth": 2,
              "marginBottom": 5,
              "marginRight": 5,
            },
            Object {
              "borderColor": "#055396",
            },
            Object {
              "backgroundColor": "#0E73CA",
            },
            undefined,
            Object {},
            Object {
              "borderBottomWidth": 0,
              "borderWidth": 0,
            },
            Object {},
            Object {},
            null,
            null,
          ],
        ]
      }
    >
      <Text
        data-class=""
        style={
          Array [
            Array [
              Object {
                "color": "#545454",
                "fontSize": 13,
                "lineHeight": 20,
              },
              Object {
                "width": "100%",
              },
              Object {
                "color": "#FFFFFF",
              },
              Array [
                undefined,
                Array [
                  undefined,
                  Array [
                    Object {
                      "fontSize": 14,
                      "paddingBottom": 7,
                      "paddingLeft": 20,
                      "paddingRight": 20,
                      "paddingTop": 7,
                      "textAlign": "center",
                    },
                    Object {},
                    Object {},
                    null,
                  ],
                ],
              ],
            ],
            Object {
              "fontFamily": "Lucida Sans",
            },
          ]
        }
      >
        Change Background Color
      </Text>
    </View>
  </View>
</View>
`;

exports[`View should render with class="Test__class".
There should have an onRef referencing to a Function. 1`] = `
<View
  class="Test__class"
  data-class=""
  onref={[Function]}
  style={
    Array [
      undefined,
      undefined,
    ]
  }
>
  <View
    accessible={true}
    isTVSelectable={true}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "opacity": 1,
        "width": "100%",
      }
    }
  >
    <View
      data-class="button button-navy"
      style={
        Array [
          undefined,
          Array [
            Object {
              "borderBottomWidth": 2,
              "marginBottom": 5,
              "marginRight": 5,
            },
            Object {
              "borderColor": "#055396",
            },
            Object {
              "backgroundColor": "#0E73CA",
            },
            undefined,
            Object {},
            Object {
              "borderBottomWidth": 0,
              "borderWidth": 0,
            },
            Object {},
            Object {},
            null,
            null,
          ],
        ]
      }
    >
      <Text
        data-class=""
        style={
          Array [
            Array [
              Object {
                "color": "#545454",
                "fontSize": 13,
                "lineHeight": 20,
              },
              Object {
                "width": "100%",
              },
              Object {
                "color": "#FFFFFF",
              },
              Array [
                undefined,
                Array [
                  undefined,
                  Array [
                    Object {
                      "fontSize": 14,
                      "paddingBottom": 7,
                      "paddingLeft": 20,
                      "paddingRight": 20,
                      "paddingTop": 7,
                      "textAlign": "center",
                    },
                    Object {},
                    Object {},
                    null,
                  ],
                ],
              ],
            ],
            Object {
              "fontFamily": "Lucida Sans",
            },
          ]
        }
      >
        Change Background Color
      </Text>
    </View>
  </View>
</View>
`;
